### PR TITLE
feat: More updates to TeslaMate

### PR DESCRIPTION
**Pull request by Megabytemb**
Added more values to be updates from MQTT.

Added a little extra logging

### CODE REVIEW

File: custom_components/tesla_custom/util.py [hunk 1]
```diff
@@ -9,3 +9,19 @@
     from homeassistant.util.ssl import client_context
 
     SSL_CONTEXT = client_context()
+
+
+def km_to_miles(odometer: float) -> float:
```
Comment (alandtse) [inline custom_components/tesla_custom/util.py:14]: We shouldn't have any conversion code. Is there no way to use HA native conversion code? Normally you just declare the native sensor units and HA figures it out.

File: custom_components/tesla_custom/teslamate.py [hunk 2]
```diff
@@ -194,6 +213,12 @@ async def async_handle_new_data(self, car: TeslaCar, msg: ReceiveMessage):
             self.update_climate_state(car, attr, cast(msg.payload))
             coordinator.async_update_listeners()
 
+        elif mqtt_attr in MAP_CHARGE_STATE:
```
Comment (alandtse) [inline custom_components/tesla_custom/teslamate.py:216]: Are there any frequent MQTT changes that we may want to explicitly mark as not being saved in the recorder?

**Review by alandtse — COMMENTED** [review]
Please address questions.

File: custom_components/tesla_custom/teslamate.py (hunk 2 above)
Comment (Megabytemb) [reply custom_components/tesla_custom/teslamate.py:216]: Not that i've noticed. When teslaMate is connected and working, MQTT updates are about 1/Second
I would think we want the end user to be able to exclude from the recorder, compared to us.

I haven't noticed any extra obseen values in my recorder

Comment (alandtse) [reply custom_components/tesla_custom/teslamate.py:234]: Sorry, are these entity attributes? I'm referring to the fact that any attribute changes will get logged in the recorder even if the main state doesn't change. This commit shows how to fix it if it does have an attribute that isn't worth recording. https://github.com/home-assistant/core/blob/3e5e937541dd368f3d70c7557ceaa7a07fd284e2/homeassistant/components/unifiprotect/recorder.py

File: custom_components/tesla_custom/teslamate.py [hunk 3]
```diff
@@ -15,18 +15,36 @@
...
 from .const import TESLAMATE_STORAGE_KEY, TESLAMATE_STORAGE_VERSION
-from .util import km_to_miles
 
 if TYPE_CHECKING:
     from . import TeslaDataUpdateCoordinator
 
 logger = logging.getLogger(__name__)
 
+
+def cast_odometer(odometer: float) -> float:
```
[the remark this answers is not in the document]
Comment (Megabytemb) [reply custom_components/tesla_custom/teslamate.py:32]: The issue is that the data Source has different values.
The Raw tesla API (and thus, what the integration and sensor code is expecting) is in Miles.
TeslaMate exposes the odometer in KMs.

In a weird way, by doing a conversion, we are actually keeping data consistent.

[the remark this answers is not in the document]
Comment (alandtse) [reply custom_components/tesla_custom/teslamate.py:32]: I may not be explaining it well. HA sensors are now supposed to just declare whatever they natively use and HA will handle all conversions. You don't have to convert any input data anymore.

**Comment by alandtse** [comment]
Also, does this cover #573

File: custom_components/tesla_custom/teslamate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Megabytemb) [reply custom_components/tesla_custom/teslamate.py:32]: I'm Aware of the Home Assistant Conversions.

The issue here is that Teslamate introduces a second source of data.
The sensors configured in the integration expect the Raw data to be in Miles.

https://github.com/alandtse/tesla/blob/437212bb9deb3de391b6114b8fd10eac00e9b8ed/custom_components/tesla_custom/sensor.py#L282

This is because the raw data provided from the Tesla API is in Miles.

However, the MQTT data from TeslaMate is in Kilometers.
https://docs.teslamate.org/docs/integrations/mqtt/

So I need to convert the TeslaMate data back to Miles so it matches the original underlying data from the Tesla API.

**Comment by Megabytemb** [comment]
> Also, does this cover #573

Latest commit fixes that issue.

I checked, all other units are the same (°C for temps, and BAR for tire pressure)

File: custom_components/tesla_custom/teslamate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (noneabove1182) [reply custom_components/tesla_custom/teslamate.py:32]: Is this not kind of expensive computationally since Teslamate is sending data every second? Would it be better to only convert Teslamate to miles if the user is using imperial, and if it's not, convert the Tesla API to metric?

[the remark this answers is not in the document]
Comment (alandtse) [reply custom_components/tesla_custom/teslamate.py:32]: > I'm Aware of the Home Assistant Conversions.
> 
> The issue here is that Teslamate introduces a second source of data.
> The sensors configured in the integration expect the Raw data to be in Miles.
> 
> https://github.com/alandtse/tesla/blob/437212bb9deb3de391b6114b8fd10eac00e9b8ed/custom_components/tesla_custom/sensor.py#L282
> 
> This is because the raw data provided from the Tesla API is in Miles.
> 
> However, the MQTT data from TeslaMate is in Kilometers.
> https://docs.teslamate.org/docs/integrations/mqtt/
> 
> So I need to convert the TeslaMate data back to Miles so it matches the original underlying data from the Tesla API.
> 

Ahh yes, you're just augmenting the original sensor and HA doesn't natively support two data sources for the same sensor. Normally they'd say have two sensors and use a template sensor to normalize the data.

[the remark this answers is not in the document]
Comment (alandtse) [reply custom_components/tesla_custom/teslamate.py:32]: HA is doing tons of conversions a second, I don't think that's an issue. I'm ok with using this to do the conversion.

File: custom_components/tesla_custom/teslamate.py (hunk 2 above)
Comment (Megabytemb) [reply custom_components/tesla_custom/teslamate.py:250]: So I looked into this.

Unifi is the only non-native component that uses that function (the rest are like Sensor, climate, select, etc...)
So it doesn't really seem like something we should use.

Most of the attributes we're pulling in from TeslaMate end up as States anyway, so Home Assistant's state handling should keep the database in check (aka, don't log if the value hasn't changed)

The only exception to this is
- charger_actual_current
- charger_voltage

And they are only populated While the vechile is charging, which would be getting database updates anyway due to the TeslaCarChargerPower sensor updating

In short, I don't think we need to do the recorder.py for this.

Comment (alandtse) [reply custom_components/tesla_custom/teslamate.py:250]: Ok, if we're creating state sensors that's fine. The concern is that any time an attribute changes on a sensor, all the data (state and attributes) are recorded in the recorder. If something changes frequently but doesn't really mean anything in the long run, you'd want to exclude that attribute from the recorder. A contrived example might be if you also added a seconds charging attribute to a battery level. Every time the seconds changed, you'd record a new entry even if the battery state may not have changed in that one second.

I thought we were creating a lot more attributes now, so I asked. If that's not the case, then that's fine.

Comment (Megabytemb) [reply custom_components/tesla_custom/teslamate.py:250]: So we're good? I don't think there's much else to add to this PR
